### DISCUSSION TITLE
Hide .glb or .gltf file on source codes

### DISCUSSION THREAD

**Discussion by sercandemirtas** [google/model-viewer#1946] [Q&A]
Hello all! 

Is it possible to hide 3d file on source codes. If you check my demo link below when you right click on page and open source code panel you can easily download my glb file from html based assets. Any idea or suggestions to hide it from users/visitors would be great. Thanks..

ps: I deployed my html, css, js based static webpage via Netlify. 

Demo Link
https://confident-lamport-09f645.netlify.app/m6.html

The paart I want to hide is **src line**
    
  class="a"
  src="/Assets/m6/m6.glb"
  alt="3D Koltuk by BaseBros"

`

### Browser Affected

- [x ] Chrome
- [x ] Edge
- [ x] Firefox
- [ x] IE
- [x ] Safari

**Comment by elalish** [comment]
Not that I'm aware of. You can look at the network tab to see what files are downloaded even if they aren't clear from your code. It's exactly like an image; you can protect it by only making a low-res version downloadable for free, or by watermarking it, or by bringing legal action against people who use it without permission. There's no equivalent to video DRM in the 3D space.

**Comment by daniel-limar2u** [comment]
Hey @sercandemirtas you can hide it in a 'post-loading' way. Here is an example: 

const observer = new MutationObserver(function (mutations) {
    if (document.contains(document.querySelector('model-viewer'))) {
      const modelviewer = document.querySelector('model-viewer')
      modelviewer.src = glb
      observer.disconnect()
    }
  })

  observer.observe(document, {
    attributes: false,
    childList: true,
    characterData: false,
    subtree: true,
  })

The src will be setted to unknown when the page is inspected. Unfortunately, will still be possible to track the file via network on devTools. Hope it will help you.

**Comment by shamsail** [comment]
1) Get file from the asset path and make a new file locally.

let file = await fetch("url/model.glb")
     .then((r) => r.blob())
     .then((blobFile) => new File([blobFile], "model.glb"));

2) create a url for local saved file
    URL.createObjectURL(file)

3) Assign this url to modelViewer.src
    modelViewer.src = URL.createObjectURL(file)
    
Hope it will work, This will not show the src path :)

**Reply by Robeater** [reply]
Hi @shamsail, 

How would I go about implementing this in my code? I tried just copy/pasting it in and the src of my models is still visible in the console. Afraid I don't have much experience with javascript.

Thanks!

![1](https://user-images.githubusercontent.com/48926304/113226943-389b4d00-9246-11eb-88b6-cb3b5fe9a99a.png)

**Comment by edaerdogann** [comment]
> Hello all!
> 
> Is it possible to hide 3d file on source codes. If you check my demo link below when you right click on page and open source code panel you can easily download my glb file from html based assets. Any idea or suggestions to hide it from users/visitors would be great. Thanks..
> 
> ps: I deployed my html, css, js based static webpage via Netlify.
> 
> Demo Link https://confident-lamport-09f645.netlify.app/m6.html
> 
> The paart I want to hide is **src line**       class="a"   src="/Assets/m6/m6.glb"   alt="3D Koltuk by BaseBros"
> 
> `
> 
> ### Browser Affected
> * [x ] Chrome
> * [x ] Edge
> * [ x] Firefox
> * [ x] IE
> * [x ] Safari



Merhaba Sercan,

Sitem için 3d görsellerimin indirilebilirliğini korumak istiyorum aynı şekilde. Profilindeki sayfayı incelediğimde görünen glb dosyası değil de başka bir dosya iniyor. Bunu nasıl yaptın acaba? Yardımcı olabilirsen çok sevinirim.

**Reply by sercandemirtas** [reply]
Eda selamlar. Testi yapmış olduğun link off durumda. Hangi linkten test ediyorsun ?

**Reply by edaerdogann** [reply]
Anasayfadaki owl için denemiştim fakat indirememişim, yanlış alarm :D Ben de aynı şekilde linki off hale getirmek istiyorum fakat yapamadım. Nasıl yaptın acaba?
